include/libdnf/repo/file_downloader.hpp

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright Contributors to the libdnf project.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2.1 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef LIBDNF_REPO_FILE_DOWNLOADER_HPP
+#define LIBDNF_REPO_FILE_DOWNLOADER_HPP
+
+#include "libdnf/common/exception.hpp"
+#include "libdnf/conf/config_main.hpp"
+#include "libdnf/repo/repo_weak.hpp"
+
+#include <memory>
+
+namespace libdnf::repo {
+
+/// Base class for file (URL) download callbacks.
+/// To implement a callback, inherit from this class and override the virtual methods.
+class FileDownloadCallbacks {
+public:
+    /// Transfer status codes.
+    enum class TransferStatus { SUCCESSFUL, ALREADYEXISTS, ERROR };
+
+    virtual ~FileDownloadCallbacks() = default;
+
+    /// End of download callback.
+    /// @param status The transfer status.
+    /// @param msg The error message in case of error.
+    /// @return TODO(lukash) uses the LrCbReturnCode enum from librepo, we should translate that.
+    virtual int end(TransferStatus status, const char * msg);
+
+    /// Download progress callback.
+    /// @param total_to_download Total number of bytes to download.
+    /// @param downloaded Number of bytes downloaded.
+    /// @return TODO(lukash) uses the LrCbReturnCode enum from librepo, we should translate that.
+    virtual int progress(double total_to_download, double downloaded);
+
+    /// Mirror failure callback.
+    /// @param msg Error message.
+    /// @param url Failed mirror URL.
+    /// @return TODO(lukash) uses the LrCbReturnCode enum from librepo, we should translate that.
+    virtual int mirror_failure(const char * msg, const char * url);
+};
```

Comment:
This callbacks interface is the same as `PackageDownloadCallbacks`, don't we want to share it?

Reply:
We want. I will use one shared code.